etl/db.py

```diff
@@ -79,19 +79,25 @@ def get_dataset_id(db_conn: MySQLdb.Connection, dataset_name: str) -> Any:
         Dataset ID.
 
     """
-    query = f"""
-        SELECT id
-        FROM datasets
-        WHERE name = '{dataset_name}'
-    """
-    with db_conn.cursor() as cursor:
-        cursor.execute(query)
-        result = cursor.fetchall()
 
-    assert len(result) == 1, f"Ambiguous or unknown dataset name '{dataset_name}'"
-    dataset_id = result[0][0]
+    def _get_dataset_id(db_conn, dataset_name):
+        query = f"""
+            SELECT id
+            FROM datasets
+            WHERE name = '{dataset_name}'
+        """
+        with db_conn.cursor() as cursor:
+            cursor.execute(query)
+            result = cursor.fetchall()
+
+        assert len(result) == 1, f"Ambiguous or unknown dataset name '{dataset_name}'"
+        dataset_id = result[0][0]
+        return dataset_id
 
-    return dataset_id
+    if db_conn is None:
+        with get_connection() as db_conn:
+            return _get_dataset_id(db_conn, dataset_name)
+    return _get_dataset_id(db_conn, dataset_name)
 
```

Comment:
If you set `db_conn` in the first line, then you won't need an inner method at all. It will look like:

```python
if db_conn is None:
    db_conn = get_connection()

# do stuff with it below
```

Reply:
oh man... I made it so complex